### ISSUE TITLE
Stripe Connect integration during account registration

### ISSUE THREAD

**Issue by DrewAPicture** [SellBirdHQ/theme#18]
[Stripe Connect](https://stripe.com/connect) is the system we will use to connect store owner accounts with our main Stripe account. This, among other things, allows us to:

- Create Stripe accounts on behalf of store owner's (also allows connecting existing accounts)
- Take a small percentage and/or flat fee of each transaction processed through stores
- See volume reports of all store owners (saves us building extensive reporting)
- Provides the means to process card payments from customers

The account registration flow will be something like this:

1. Register for account on sellbird.com (goes through RCP)
2. Subsite is created for account owner in our multisite network
3. Account owner lands in their dashboard and are given the option to create products and set up their various store settings (description, logo, etc)
4. A banner notice is shown to indicate their store is in "pending" or "test" mode and they need to go live before they can start selling
5. To go live, account owner clicks on a button that sends them through the Stripe Connect oauth flow
6. Once oauth is complete, account owner lands back on sellbird.com and their account is switched to live mode

With Stripe Connect, we actually have three different types of accounts we can use:

- Standard
- Express
- Custom

Each has their own pros and cons and we'll need to decide exactly which one we want to use. Please review Stripe's description of them: https://stripe.com/docs/connect/accounts

While choosing the account type we want to use, we also need to consider the pricing differences: https://stripe.com/us/connect/pricing

[Connect documentation](https://stripe.com/docs/connect).

Previously: https://github.com/SellBirdHQ/sellbird/issues/20

**Comment by DrewAPicture** [comment]
Reading up a little bit on the Stripe site, it seems like the best option is to use Express accounts, up until you read that they are effectively U.S. only. Which leads us to Custom accounts, where you have to do a lot more of the legwork in gathering information and management (because custom accounts don't have access to the Stripe dashboard).

**Comment by DrewAPicture** [comment]
Standard is also an option and it seems to offer some of the same benefits custom accounts would (any country sellers can have an account for) but with the caveat that they're essentially connecting an existing account to the platform.

According to the [Creating Charges](https://stripe.com/docs/connect/charges) docs, there might be some reasonably good options for handling the platform fees (percentage). Currently my vote is on Standard accounts.

**Comment by pippinsplugins** [comment]
Do you see any reason we couldn't use Express for US customers and Standard for non-US customers? We could also launch with US-only initially if needed.

**Comment by DrewAPicture** [comment]
I got so used to just committing to master that I spaced on creating an issue branch haha. See https://github.com/SellBirdHQ/theme/compare/9b2f1db1...master

**Comment by DrewAPicture** [comment]
Reverted the stuff in `master` in favor of starting a PR: #28

**Comment by DrewAPicture** [comment]
So, this is ready for testing. As noted in Slack, due the limitations on development vs production domains, when you click the link in the Dashboard to connect to Stripe, and then return back to the store, you'll have to manually change the domain to .com from .dev. Stripe won't actually record a connected account for test accounts, but if the URL on the return contains a `code` key and value, it's a "successful" connection.

**Comment by pippinsplugins** [comment]
@DrewAPicture we'll need to go ahead and go a bit further here.

After the connection to Stripe is made, we need to insert the account ID into the customer's account record so that we can then instruct Stripe which account to use when processing purchases through a user's store.

**Comment by DrewAPicture** [comment]
I've added the logic for the second part of the handshake that retrieves the `stripe_user_id` and `refresh_token` values and stores them in the options table for the current site. This appears to have been missing link for connected dev accounts not showing up in the Stripe Connect dashboard.

To test, you'll need to pull down an updated copy of the `includes/class-secrets.php` file in the theme, which contains the recently-added API keys needed to send the secondary request.

**Comment by DrewAPicture** [comment]
Also worth noting that I've removed the logic for activating a site following connection to Stripe. I think "going live" should be a conscious decision removed from the process of connecting to Stripe, which is really just one step in getting set up.

**Comment by pippinsplugins** [comment]
@DrewAPicture Either I've missed something or this isn't working.

I just did a test registration and after the account was created, I was shown this:
![screen shot 2017-09-07 at 12 43 39 pm](https://user-images.githubusercontent.com/1034109/30177122-62aada8a-93ca-11e7-9746-1b871f249d6c.png)

The site was created but was not attributed to my newly created account. Not connection to Stripe was ever made either.

Am I missing something?

**Comment by DrewAPicture** [comment]
Let's close this in favor of a new issue. It worked when it was merged, so something else has to have changed to cause it not to now. Also, can you further outline your steps to reproduce in the new issue?

**Comment by pippinsplugins** [comment]
@DrewAPicture my steps were super simple: register a brand new account at /register. Doing that will replicate the problem.

**Comment by pippinsplugins** [comment]
Re-opening this since it isn't fully functional yet. Connecting to Stripe during account registration is currently not happening as it should.

To replicate the issue, visit /register and create a new account. At no point (currently) is the customer sent through the Connect flow nor are they shown any way to initiate the connection.

**Comment by DrewAPicture** [comment]
Started a new PR (WIP): #63

**Comment by DrewAPicture** [comment]
@pippinsplugins I believe #63 is ready for testing, per your checklist:

* [x] The process during account registration where the customer is sent to Stripe, either as part of registration or as a step after registration
* [x] The complete authorization process with Stripe
* [x] The data connection that allows purchases to be made on the user's EDD store through the connected Stripe account. To do this, you will need to tie into the `edd-stripe` gateway and set the necessary API parameters so that the Stripe API uses  the connected account

Connecting a new store to Stripe is still contingent on the membership being active because the site has to exist to store options in it ;)

Most of what's being added in this PR has to do with improving microcopy and other messaging to better indicate what's needed to proceed (both in the UI and in the code).

**To test with a completely new store:**

1. Visit /register and register a new membership
2. Once registration is complete you should be redirect to /dashboard and logged-in
2b. If your RCP settings dictate that members need to be manually activated, you should see a message on /dashboard indicating that the store is still pending activation.
3. Once the membership is active, you should see a link on /dashboard to connect to Stripe.
4. Clicking the "Set Up Stripe" link should take you to Stripe to set up an account.
5. For testing purposes, simply clicking the "Skip this account form" link at the top of the external page will suffice to initiate the handshake
6. Once you've clicked "Skip this account form", you should be redirected back to /dashboard and presented with a message indicating Stripe is now connected.

To double check that things are on the up-and-up, I recommend running the following to confirm the option was stored in the second part of the Stripe handshake:
```sh
wp option get 'stripe_store_data' --url=whatever.sellbird.dev
```

**To test with an existing membership**, skip to step 3 above and proceed through. To re-test an existing membership that's already been connected, the easiest way to do that is to simply delete the option key:
```sh
wp option delete 'stripe_store_data' --url=whatever.sellbird.dev
```
If for whatever reason, the testing steps above don't seem to be working, check your error log (assuming you have `WP_DEBUG` enabled. I added some error logging for the Stripe handshake request, and there might be some useful information in the log explaining what's going on.

Screenshots:
**Inactive member notice:**
![Inactive](https://cl.ly/2H1E3C0S0745/Screen%20Shot%202017-09-29%20at%2012.31.11%20AM.png)

**Active member, Stripe not connected:**
![Not connected](https://cl.ly/2X1Q411X2s0R/Screen%20Shot%202017-09-29%20at%2012.30.26%20AM.png)

**Active member, Stripe connected:**
![Connected](https://cl.ly/2Q241N3B0Q1R/Screen%20Shot%202017-09-29%20at%2012.26.41%20AM.png)

**Comment by pippinsplugins** [comment]
Beautiful, I will try and test today!

**Comment by amdrew** [comment]
@DrewAPicture Registered a new site via `/register`. Once redirected to the dashboard I clicked the link "Set up Stripe", and the page just reloads. The link goes to nowhere (href is empty).

**Comment by DrewAPicture** [comment]
Sorry, I think I wiped out the `Secrets` class from `includes/` when I fixed the server to be able to pull from GitHub. Try it now.

**Comment by amdrew** [comment]
@DrewAPicture 

I'm back on the dashboard after skipping the form, no message though.

>  "Once you've clicked "Skip this account form", you should be redirected back to /dashboard and presented with a message indicating Stripe is now connected."

**Comment by amdrew** [comment]
Oh, it put me on my local site?

**Comment by amdrew** [comment]
Ok, changed .dev to .com, now I see it 

<img width="863" alt="screenshot 2017-10-04 14 45 56" src="https://user-images.githubusercontent.com/52581/31156792-be641faa-a912-11e7-9efe-e1d67d4c65b8.png">

**Comment by DrewAPicture** [comment]
And just to note, the latest commit (in conjunction with a setting change on Stripe) now makes the Connect link work without having to manually update the domain on the return.

**Comment by pippinsplugins** [comment]
Excellent.

@DrewAPicture ready to merge into master?

**Comment by DrewAPicture** [comment]
Yes!
